src/themis/sym_enc_message.c

```diff
@@ -181,7 +181,7 @@ themis_status_t themis_auth_sym_encrypt_message(const uint8_t* key,
 						 size_t* encrypted_message_length){
   uint8_t key_[THEMIS_AUTH_SYM_KEY_LENGTH/8];
   THEMIS_CHECK_PARAM(message!=NULL && message_length!=0);
-  THEMIS_STATUS_CHECK(themis_sym_kdf(key,key_length, THEMIS_SYM_KDF_KEY_LABEL, (uint8_t*)(&message_length), sizeof(message_length), in_context, in_context_length, key_, sizeof(key_)),THEMIS_SUCCESS);
+  THEMIS_STATUS_CHECK(themis_sym_kdf(key,key_length, THEMIS_SYM_KDF_KEY_LABEL, (uint8_t*)(&message_length), sizeof(uint32_t), in_context, in_context_length, key_, sizeof(key_)),THEMIS_SUCCESS);
```

Comment:
agree with you

unfortunately, we can't easily declare `message_length` as `uint32_t` because it requires changing function signatures all over the project (changing params from `size_t` to `uint32_t`), which requires updating most of the wrappers. (pile of work, many mistakes :doge:)

from another perspective, current code already uses `uint32_t` cast in some places like
https://github.com/cossacklabs/themis/blob/master/src/themis/sym_enc_message.c#L166 so I would say that changing `message_length` type in future is *not an easy task* anyway

Reply:
This just creates a potential landmine for the future (and maybe even a security incident). If it is possible, better change it now than suffer the consequences in the future.

Reply:
I'm on your side totally.

However i'm trying to fix 3yrs old mistake of using `sizeof(size_t)` without need of re-writing half of Themis core and introducing more mistakes. Believe me, first thing I tried was changing a type of `message_length`, but I faced tons of changed code, including secure message, secure session, tests, wrappers, everywhere.

Can we define `uint32_t` somewhere in this file and use define instead of actual type? :)

Reply:
The point is this is used to calculate buffer sizes, therefore enforces buffer safety. Our primary goal is secure code and this type of fix makes it worse in that regard. Even if it is much more work - it is justified IMO. Otherwise we might be breaking a "security promise" for Themis users.

`sizeof(type)` in buffer length calculation and using an actual value of a different type is just too typical for a potential security issue we should be making in a security oriented project.

Reply:
You're absolutely right in terms of security and 'best practices' perspective.

We will update Themis core to use `uint32_t` or even `uint64_t` for message calculations when we will have automated integration testing between platforms, wrappers and core versions. Until then we cannot guarantee that significant code changes won't break core functionality. 

Unfortunately the code already has force casts like this

```
const size_t message_length
...
hdr->message_length=(uint32_t)message_length;
```

I don't see how adding
```
size_t* encrypted_message_length
...
sizeof((uint32_t)(encrypted_message_length))
```

is better than `sizeof(uint32_t)`.

Taking into account limited resources and severity of the problem we're trying to solve, I believe that this temporary solution works before proper refactoring. We already working on integration test suit.

Can you please explain how using `sizeof(uint32_t)` decreases current security guarantees, except the point that we will forget to change this code during future refactoring.

Reply:
> Unfortunately the code already has force casts like this

That troubles me as well: casting up is fine, casting down usually creates issues (unless you're sure they are not)

> Can you please explain how using sizeof(uint32_t) decreases current security guarantees

The point is in "forgetting to change" - for example you have code which serialises secure cell structure and you usually do it by iterating each variable usually, something like
```
uint8_t *p = malloc(hdr + len);
memcpy(p, &hdr);
p += sizeof(hdr);
memcpy(p, &len);
p += sizeof(len);
...
```

there are two potential problems here if you get incorrect size forgetting to update sizeof(len) with proper type:
* if you update it in serialisation code, but not in the malloc code - it will allocate smaller buffer and you'll write data past the buffer boundary
* even if you update it in both places, but actual length is 64 bit, but we just use sizeof(uint32_t) as a storage buffer (cast down) - this code will break on big-endian systems (which are still common nowadays), because it will write zeroes at best and garbage at worst

Now these type of errors are very easy to make, they are very hard to spot during code review, they will likely not be caught by automatic test suite.

Reply:
Thank you, got it!

How about next plan:
- merge fix as is;
- document these problems (down-casting and `sizeof(uint32_t)`) as places of potential risks (add "warning-note-for-future-us" inside code);
- check code with static analyzers (afaik we are doing this manually time to time, not sure whether analyzers trigger warnings about down-casting);
- finish automated between-platforms integration tests (we are working on this right now, any ideas how to test on big-endian systems?);
- create refactoring plan, paying attention on risky places, change `size_t` variable to `uint32_t` (`uint64_t`? whatever we decide is better);
- refactor core and wrappers, keeping an eye on compatibility.

I believe that your suggestion #270 to update C code style will help us a lot!

What do you think?

Reply:
Probably adding a `TODO` in the comment above these "tricky lines" will suffice for now.

Reply:
done!